### PR TITLE
Shooter To Main

### PR DESCRIPTION
**Pull request by Mason-Beard**

### CODE REVIEW

File: src/main/java/frc/robot/RobotContainer.java [hunk 1]
```diff
@@ -83,14 +88,14 @@ public RobotContainer() {
...
-      m_visualizer.registerShooter(m_shooter);
-      m_visualizer.registerAmpShooter(m_ampShooter);
-      m_visualizer.registerArm(m_arm);
-      m_visualizer.registerClimber(m_climber);
-      //    m_visualizer.registerVision(m_vision);
-      m_visualizer.registerLedSubsystem(m_led);
+      // m_visualizer = new SuperStructureVisualizer();
+      // m_visualizer.registerIntake(m_intake);
+      // m_visualizer.registerShooter(m_shooter);
+      // m_visualizer.registerAmpShooter(m_ampShooter);
```
Comment (SherazF) [inline src/main/java/frc/robot/RobotContainer.java:94]: Uncomment please

File: src/main/java/frc/robot/RobotContainer.java [hunk 2]
```diff
@@ -242,8 +253,8 @@ public void periodic() {
     //             estimatedRobotPose.estimatedPose.toPose2d(),
     // estimatedRobotPose.timestampSeconds));
 
-    m_fieldSim.periodic();
-    if (m_visualizer != null) m_visualizer.periodic();
+    //   m_fieldSim.periodic();
```
Comment (SherazF) [inline src/main/java/frc/robot/RobotContainer.java:256]: Uncomment please

File: src/main/java/frc/robot/commands/amp/AutoSetAmpSpeed.java [hunk 3]
```diff
@@ -1,42 +1,42 @@
...
 
-//   // Called every time the scheduler runs while the command is scheduled.
+  // Called every time the scheduler runs while the command is scheduled.
 
-//   // Called once the command ends or is interrupted.
-//   @Override
-//   public void end(boolean interrupted) {}
+  // Called once the command ends or is interrupted.
+  @Override
+  public void end(boolean interrupted) {}
```
Comment (SherazF) [inline src/main/java/frc/robot/commands/amp/AutoSetAmpSpeed.java:35]: Fix this

File: src/main/java/frc/robot/commands/amp/AutoSetAmpSpeed.java [hunk 4]
```diff
@@ -1,42 +1,42 @@
...
-//   // Returns true when the command should end.
-//   @Override
-//   public boolean isFinished() {
-//     return true;
-//   }
-// }
+  // Returns true when the command should end.
+  @Override
+  public boolean isFinished() {
+    return true;
```
Comment (SherazF) [inline src/main/java/frc/robot/commands/amp/AutoSetAmpSpeed.java:40]: If it ends this should be false

File: src/main/java/frc/robot/commands/shooter/AutoSetPercentSetpoint.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
+// Copyright (c) FIRST and other WPILib contributors.
```
Comment (SherazF) [inline src/main/java/frc/robot/commands/shooter/AutoSetPercentSetpoint.java:1]: Delete this, not needed

File: src/main/java/frc/robot/commands/shooter/SetAndHoldPercentSetpoint.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
+// Copyright (c) FIRST and other WPILib contributors.
```
Comment (SherazF) [inline src/main/java/frc/robot/commands/shooter/SetAndHoldPercentSetpoint.java:1]: Delete this, not needed

File: src/main/java/frc/robot/commands/shooter/SetAndHoldRPMSetpoint.java [hunk 7]
```diff
@@ -23,15 +23,16 @@ public void initialize() {}
...
   }
 
   // Called every time the scheduler runs while the command is scheduled.
 
   // Called once the command ends or is interrupted.
   @Override
   public void end(boolean interrupted) {
-    m_shooter.setRPM(0);
+
+    m_shooter.setRpmOutput(0);
```
Comment (SherazF) [inline src/main/java/frc/robot/commands/shooter/SetAndHoldRPMSetpoint.java:35]: Fix naming on this command.

File: src/main/java/frc/robot/constants/CAN.java [hunk 8]
```diff
@@ -23,6 +23,7 @@ public final class CAN {
 
   public static final int intakeMotor1 = 30;
   public static final int intakeMotor2 = 31;
+  public static final int uptakeMotor = 35;
```
Comment (SherazF) [inline src/main/java/frc/robot/constants/CAN.java:26]: Let's rename this to Amp Motor

File: src/main/java/frc/robot/constants/FLYWHEEL.java [hunk 9]
```diff
@@ -3,12 +3,25 @@
...
+
+  public static final double kP = 10;
+  public static final double kI = 0.0;
+  public static final double kD = 1;
+
+  public static final double kS = 0.0;
+  public static final double kV = 0.0;
+  public static final double kA = 0.0;
+
+  public static final double shooterangle = 35;
```
Comment (SherazF) [inline src/main/java/frc/robot/constants/FLYWHEEL.java:19]: Usually this should be in radian. Use Unit.DegreestoRadians(35) here for functionality

File: src/main/java/frc/robot/constants/FLYWHEEL.java [hunk 10]
```diff
@@ -35,5 +48,39 @@ public enum FLYWHEEL_STATE {
     public double get() {
       return value;
     }
+
+    public Object getPercentOutput() {
```
Comment (SherazF) [inline src/main/java/frc/robot/constants/FLYWHEEL.java:52]: What is this? Can we confirm that this is needed ? Prob delete?

File: src/main/java/frc/robot/subsystems/AmpShooter.java [hunk 11]
```diff
@@ -29,20 +30,20 @@ public void setPercentOutput(double m_speed) {
...
   }
 
   private void updateShuffleboard() {
-    //    SmartDashboard.putNumber("ampShooterRPM", this.getRPM());
+    SmartDashboard.putNumber("ampShooterPercent", this.getPercentOutput());
   }
 
   private void updateLog() {
-    Logger.recordOutput("AmpShooter/RPM", getRPM());
+    Logger.recordOutput("AmpShooter/Velocity", getVelocity());
```
Comment (SherazF) [inline src/main/java/frc/robot/subsystems/AmpShooter.java:46]: Make sure that we have the RPM logged

File: src/main/java/frc/robot/constants/FLYWHEEL.java
Comment (SherazF) [inline src/main/java/frc/robot/constants/FLYWHEEL.java:1]: Rename this to Shooter.Java file

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 12]
```diff
@@ -5,51 +5,187 @@
...
+  private boolean m_testMode = false;
+  private double m_headingOffset;
   private double flywheelRPMRatio = 1.0;
+  private double m_percentOutput;
+
+  private final TalonFX[] m_shooterMotors = {
+    new TalonFX(CAN.flywheel1), new TalonFX(CAN.flywheel2) // Flywheel[0] is bottom
+  };
+
+  private double flywheelPercentRatio = 1.0;
```
Comment (SherazF) [inline src/main/java/frc/robot/subsystems/Shooter.java:33]: Not sure what this is. Please name more specifically or delete

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 13]
```diff
@@ -5,51 +5,187 @@
...
-    TalonFXConfiguration config = new TalonFXConfiguration();
-    flywheelmotor1.getConfigurator().apply(config);
-    flywheelmotor1.setInverted(true);
-    // flywheel motor 2
-    flywheelmotor2.getConfigurator().apply(config);
-    flywheelmotor2.setInverted(false);
+    m_shooterMotors[1].setControl(new Follower(m_shooterMotors[0].getDeviceID(), true));
   }
 
   // values that we set
```
Comment (SherazF) [inline src/main/java/frc/robot/subsystems/Shooter.java:71]: Make comment above this function: Only use when absolutely needed. For every other case use RPM.

**Review by SherazF — CHANGES_REQUESTED** [review]
I have left comments on everything that needs to be changed. Good work thus far!

File: src/main/java/frc/robot/constants/SHOOTER.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    WAIT(double value) {
+      this.value = value;
+    }
+
+    public double get() {
+      return value;
+    }
+  }
+
+  public enum FLYWHEEL_STATE {
```
Comment (jonathandao0) [inline src/main/java/frc/robot/constants/SHOOTER.java:38]: Change this to SETPOINT and make the values all in RPM. Update all commands that use this to use the appropriate value

File: src/main/java/frc/robot/commands/shooter/AutoSetPercentSetpoint.java [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+
+import edu.wpi.first.wpilibj2.command.Command;
+import frc.robot.constants.SHOOTER.FLYWHEEL_STATE;
+import frc.robot.subsystems.Shooter;
+
+public class AutoSetPercentSetpoint extends Command {
+  Shooter m_shooter;
+  FLYWHEEL_STATE m_state;
+
+  public AutoSetPercentSetpoint(Shooter shooter, FLYWHEEL_STATE state) {
```
Comment (jonathandao0) [inline src/main/java/frc/robot/commands/shooter/AutoSetPercentSetpoint.java:11]: If this is just using percent output, use a double and not STATE/SETPOINT

File: src/main/java/frc/robot/commands/shooter/AutoSetRPMSetpoint.java
Comment (jonathandao0) [inline src/main/java/frc/robot/commands/shooter/AutoSetRPMSetpoint.java:1]: Have it use a double argument instead of FLYWHEEL_STATE/SETPOINT

**Review by jonathandao0 — CHANGES_REQUESTED** [review]

**Review by jonathandao0 — APPROVED** [review]

**Review by jonathandao0 — APPROVED** [review]

**Pull request merged**